File: java/leetcode/backtracking/subsets_ii.java
/**
 * 90. Subsets II
 * Given a collection of integers that might contain duplicates, nums, return all possible subsets (the power set).

Note: The solution set must not contain duplicate subsets.

Example:

Input: [1,2,2]
Output:
[
  [2],
  [1],
  [1,2,2],
  [2,2],
  [1,2],
  []
]
 */


import java.util.List;
import java.util.ArrayList;
import java.util.Arrays;

public class subsets_ii {

    public  List<List<Integer>> subsetsWithDup(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> subsets = new ArrayList<>();
        List<Integer> tempSubset = new ArrayList<>();
        boolean[] hasVisited = new boolean[nums.length];
        for (int size = 0; size <= nums.length; size++) {
            backtracking(0, tempSubset, subsets, hasVisited, size, nums);
        }
        return subsets;
    }

    private void backtracking(int start, List<Integer> tempSubset, List<List<Integer>> subsets, 
        boolean[] hasVisited, final int size, final int[] nums) {

        if (tempSubset.size() == size) {
            subsets.add(new ArrayList<>(tempSubset));
            return;
        }
        for (int i = start; i < nums.length; i++) {
            if (i != 0 && nums[i] == nums[i-1] && !hasVisited[i-1]) {
                continue;
            }
            tempSubset.add(nums[i]);
            hasVisited[i] = true;
            backtracking(i+1, tempSubset, subsets, hasVisited, size, nums);
            hasVisited[i] = false;
            tempSubset.remove(tempSubset.size()-1);
        }
    }
}

File: golang/leetcode/backtracking/subsets_ii.go
package backtracking

import "sort"

// 90. Subsets II

func subsetsWithDup(nums []int) [][]int {
	sort.Ints(nums)
	results := make([][]int, 0)
	temp := make([]int, 0)
	visited := make([]bool, len(nums))
	for size := 0; size <= len(nums); size++ {
		subsetsTrack2(0, temp, &results, visited, size, nums)
	}
	return results
}

func subsetsTrack2(start int, temp []int, results *[][]int, visited []bool, size int, nums []int) {
	if len(temp) == size {
		combo := make([]int, len(temp))
		copy(combo, temp)
		*results = append(*results, combo)
		return
	}
	for i := start; i < len(nums); i++ {
		if i != 0 && nums[i] == nums[i-1] && !visited[i-1] {
			continue
		}
		temp = append(temp, nums[i])
		visited[i] = true
		subsetsTrack2(i+1, temp, results, visited, size, nums)
		visited[i] = false
		temp = temp[:len(temp)-1]
	}
}
